infrahub_sdk/transforms.py

```diff
@@ -72,17 +90,40 @@ def branch_name(self) -> str:
     def transform(self, data: dict) -> Any:
         pass
 
-    async def collect_data(self) -> dict:
+    async def collect_data(self, variables: Optional[dict] = None) -> dict:
         """Query the result of the GraphQL Query defined in self.query and return the result"""
 
-        return await self.client.query_gql_query(name=self.query, branch_name=self.branch_name)
+        if not variables:
+            variables = {}
 
-    async def run(self, data: Optional[dict] = None) -> Any:
+        # Try getting data from stored graphql query endpoint
+        try:
+            return await self.client.query_gql_query(name=self.query, branch_name=self.branch_name, variables=variables)
+        # If we run into an error, the stored graphql query may not exist. Try to query the GraphQL API directly instead.
+        except httpx.HTTPStatusError:
+            if not self.repository_config:
+                raise
+            query_str = self.repository_config.get_query(name=self.query).load_query()
+            return await self.client.execute_graphql(query=query_str, variables=variables, branch_name=self.branch_name)
```

Comment:
I'd like to revert to the old behaviour here so that the focus of this PR is only to add the missing `InfrahubClient` that was the target of #8.

While I can see that the collect_data is missing some information as it is now I'm not sure that we should add a `InfrahubRepositoryConfig` object to the transform itself in order to load a missing query. I think we should handle all that in a different way. I'm curious to the root cause of this if you had problems running with a local query if it had to do with the fact that the query wasn't defined within your `.infrahub.yml` file? This is my assumption based on the `repository_config.queries.append(query_config_obj)` could this be correct?

Reply:
The query was indeed not in my infrahub.yml file (hence the loading of the query in L#352-354 of `cli_commands.py`, which can be removed with this in mind) -- but that isn't the root of the issue with running the transform here.

The InfrahubTransform object in stable only has access to the query name, not to the query string. The query name is able to be used in the front end because a stored query at the path `f"{self.address}/api/query/{name}" ` pre-exists in infrahub which returns data, so we don't actually need the query string to make the query, just the name of the query. I think this stored query path is created as part of syncing a repo (though I'm not sure). This is not the case when the `infrahubctl transform` command is run, because the stored query path does not pre-exist in infrahub (unless I'm doing something wrong here). This in mind, we need the full query string to execute the query with `infrahubctl transform`, not just the query name.

```
HTTP communication failure: Client error '403 Forbidden' for url 'http://localhost:8000/api/query/tags_query?branch=main&update_group=false'
For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403 on POST to 
http://localhost:8000/api/query/tags_query?branch=main&update_group=false
```

I can add a client object to the transform (as per #8), but I'll have to call the client object's `.execute_graphql` method directly with the query string in order to acquire data, circumventing the transform's run method for data acquisition. I can then run the transform with that data. Is this what you're thinking would be best to resolve #8 with a more modest touch?

Reply:
I updated per my comment there. If this isn't what you're thinking (or I'm missing something) LMK and I can modify.